### PR TITLE
Add custom rbac verbs to allow modification of `NamespacedCloudProfile.Spec.{Kubernetes,MachineImages}`

### PR DESCRIPTION
**Pull request by LucaBernstein**
**How to categorize this PR?**

/area usability
/area scalability
/kind api-change
/kind enhancement

**What this PR does / why we need it**:
- Grant project members access to `NamespacedCloudProfile`s, with different capabilities depending on their role within the project.
- Introduce custom rbac verbs `modify-spec-kubernetes` and `modify-spec-machineimages` to allow modifications on a `NamespacedCloudProfile` in the `kubernetes` and `machineImages` sections. Typically to be used by operators.
- ~~Reorder `Shoot` api response so that `cloudProfile` and `cloudProfileName` are placed next to each other. This has a visible impact especially on the dashboard in the `Shoot` "YAML" page view.~~

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/9504.

**Special notes for your reviewer**:
/cc @timuthy 

**Release note**:

```noteworthy operator
Allow project users to read `NamespacedCloudProfile`s and for project admins to make adjustments to machine types and volume types.
```

```feature operator
Introduce custom RBAC verbs to allow for modification of `.spec.{kubernetes,machineImages}` in `NamespacedCloudProfile`s.
```

### CODE REVIEW

File: plugin/pkg/global/customverbauthorizer/admission.go [hunk 1]
```diff
@@ -31,6 +31,13 @@ const (
 	// CustomVerbProjectManageMembers is a constant for the custom verb that allows to manage human users or
 	// groups subjects in the `.spec.members` field in `Project` resources.
 	CustomVerbProjectManageMembers = "manage-members"
+
+	// CustomVerbNscpflModifyKubernetes is a constant for the custom verb that allows modifying the
+	// `.spec.kubernetes` field in `NamespacedCloudProfile` resources.
+	CustomVerbNscpflModifyKubernetes = "modify-spec-kubernetes"
+	// CustomVerbNscpflModifyMachineImages is a constant for the custom verb that allows modifying the
+	// `.spec.machineImages` field in `NamespacedCloudProfile` resources.
+	CustomVerbNscpflModifyMachineImages = "modify-spec-machineimages"
```
Comment (rfranzke) [inline plugin/pkg/global/customverbauthorizer/admission.go]: Do not use uncommon abbreviations like` Nscpfl`. Either remove this part, or call it `NamespacedCloudProfile`.

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 2]
```diff
@@ -63,6 +63,9 @@ type ShootSpec struct {
 	// Addons contains information about enabled/disabled addons and their configuration.
 	// +optional
 	Addons *Addons `json:"addons,omitempty" protobuf:"bytes,1,opt,name=addons"`
+	// CloudProfile contains a reference to a CloudProfile or a NamespacedCloudProfile.
+	// +optional
+	CloudProfile *CloudProfileReference `json:"cloudProfile,omitempty" protobuf:"bytes,22,opt,name=cloudProfile"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go]: Please move this down again. While I understand the motivation, the protobuf number ordering gets messed up this way.

File: plugin/pkg/global/customverbauthorizer/admission.go [hunk 3]
```diff
@@ -110,6 +119,42 @@ func (c *CustomVerbAuthorizer) admitProjects(ctx context.Context, a admission.At
...
+	}
+
+	if a.GetOperation() == admission.Update {
+		oldObj, ok = a.GetOldObject().(*core.NamespacedCloudProfile)
+		if !ok {
+			return apierrors.NewBadRequest("could not convert old resource into NamespacedCloudProfile object")
+		}
+	}
+
+	if mustCheckNscpflKubernetes(oldObj.Spec.Kubernetes, obj.Spec.Kubernetes) {
```
Comment (rfranzke) [inline plugin/pkg/global/customverbauthorizer/admission.go]: same here and other occurrences

**Review by rfranzke — COMMENTED** [review]
Can you document this feature somewhere, please?

File: pkg/apis/core/v1beta1/types_shoot.go (hunk 2 above)
Comment (LucaBernstein) [reply pkg/apis/core/v1beta1/types_shoot.go]: I have dropped the respective commit. Either dashboard needs to sort then or we wait until the Shoot api is promoted to stable v1.

**Comment by LucaBernstein** [comment]
> Can you document this feature somewhere, please?

Thank you for your review.
I improved the variable namings and added some documentation where I saw fit. (ref https://github.com/gardener/gardener/commit/a0307ecb7e82ba90f586cce7c8da5ecb98ec171f)

File: charts/gardener/controlplane/charts/application/templates/rbac-user.yaml [hunk 4]
```diff
@@ -502,6 +501,16 @@ rules:
   - shoots/viewerkubeconfig
   verbs:
   - create
+- apiGroups:
+  - core.gardener.cloud
+  resources:
+  - namespacedcloudprofiles
+  verbs:
+  - get
```
Comment (timuthy) [inline charts/gardener/controlplane/charts/application/templates/rbac-user.yaml:509]:
```suggestion
  - create
  - delete
  - get
```

File: pkg/component/garden/system/virtual/virtual.go [hunk 5]
```diff
@@ -439,6 +440,11 @@ func (g *gardenSystem) computeResourcesData() (map[string][]byte, error) {
 					},
 					Verbs: []string{"create"},
 				},
+				{
+					APIGroups: []string{gardencorev1beta1.GroupName},
+					Resources: []string{"namespacedcloudprofiles"},
+					Verbs:     []string{"get", "list", "watch", "patch", "update"},
```
Comment (timuthy) [inline pkg/component/garden/system/virtual/virtual.go]:
```suggestion
					Verbs:     []string{"create", "delete", "get", "list", "watch", "patch", "update"},
```

**Review by timuthy — CHANGES_REQUESTED** [review]
Comments as discussed.
Very nice PR, thanks for the changes 🙂

File: docs/concepts/apiserver.md [hunk 6]
```diff
@@ -35,6 +35,10 @@ At the moment, only the necessary APIs and validations exist to allow for extens
 When a shoot is created or updated, the cloudprofile reference can be set to point to a directly descendant `NamespacedCloudProfile`.
 Updates from one `CloudProfile` to another `CloudProfile` or from one `NamespacedCloudProfile` to another `NamespacedCloudProfile` or even to another `CloudProfile` are not allowed.
 
+Project viewers have the permission to see `NamespacedCloudProfile`s associated with a particular project.
+Project members can edit the `NamespacedCloudProfile` spec, except for the special fields `.spec.kubernetes` and `.spec.machineImages`.
```
Comment (timuthy) [inline docs/concepts/apiserver.md]:
```suggestion
Project members can create, edit or delete `NamespacedCloudProfile`s, except for the special fields `.spec.kubernetes` and `.spec.machineImages`.
```

File: docs/concepts/apiserver.md [hunk 7]
```diff
@@ -35,6 +35,10 @@ At the moment, only the necessary APIs and validations exist to allow for extens
...
+In order to make changes to these special fields, a user needs to be granted the custom RBAC verbs `modify-spec-kubernetes` and `modify-spec-machineimages` respectively.
```
Comment (timuthy) [inline docs/concepts/apiserver.md]:
```suggestion
In order to make changes to these special fields, a user needs to be granted the custom RBAC verbs `modify-spec-kubernetes` and `modify-spec-machineimages` respectively, which is typically only granted to landscape operators.
```

File: test/integration/controllermanager/project/project/project_test.go [hunk 8]
```diff
@@ -536,5 +537,250 @@ var _ = Describe("Project controller tests", func() {
...
+					Versions: []gardencorev1beta1.ExpirableVersion{{Version: "1.25.1", ExpirationDate: futureExpirationDate}},
+				}
+				namespacedCloudProfile.Spec.MachineImages = []gardencorev1beta1.MachineImage{
+					{Name: "some-OS", Versions: []gardencorev1beta1.MachineImageVersion{
+						{ExpirableVersion: gardencorev1beta1.ExpirableVersion{Version: "1.1.1", ExpirationDate: futureExpirationDate}},
+					}},
+				}
+				Eventually(func(g Gomega) {
+					g.Expect(testUserClient.Update(ctx, namespacedCloudProfile)).To(Succeed())
+				}).Should(Succeed())
```
Comment (timuthy) [inline test/integration/controllermanager/project/project/project_test.go]:
```suggestion
				Eventually(func() error {
					return testUserClient.Update(ctx, namespacedCloudProfile)
				}).Should(Succeed())
```

Nit, so you don't need double `Succeed` calls. Same for other occurrences.

File: test/integration/controllermanager/project/project/project_test.go [hunk 9]
```diff
@@ -536,5 +537,250 @@ var _ = Describe("Project controller tests", func() {
...
+
+				By("Ensure admin without proper role can update NamespacedCloudProfile.Spec.{MachineTypes,VolumeTypes}")
+				namespacedCloudProfile.Spec.MachineTypes = []gardencorev1beta1.MachineType{
+					{Name: "test-machine-type", CPU: resource.MustParse("2"), Memory: resource.MustParse("1G")},
+				}
+				namespacedCloudProfile.Spec.VolumeTypes = []gardencorev1beta1.VolumeType{
+					{Name: "test-volume-type", Class: "standard"},
+				}
+				Expect(testUserClient.Update(ctx, namespacedCloudProfile)).To(Succeed())
```
Comment (timuthy) [inline test/integration/controllermanager/project/project/project_test.go:711]: Unlike for other `update` calls, `Eventually` is not used here. Is that intended?

**Review by timuthy — CHANGES_REQUESTED** [review]

File: test/integration/controllermanager/project/project/project_test.go (hunk 9 above)
Comment (LucaBernstein) [reply test/integration/controllermanager/project/project/project_test.go:711]: If I ensured test user access to the NamespacedCloudProfile earlier already, I omit the `Eventually`, as I really expect it to still work (from authorization point of view).

**Review by timuthy — COMMENTED** [review]
/lgtm

File: test/integration/controllermanager/project/project/project_suite_test.go [hunk 10]
```diff
@@ -133,6 +138,22 @@ var _ = BeforeSuite(func() {
...
+			ConcurrentSyncs: ptr.To(5),
+		},
+	}).AddToManager(mgr)).To(Succeed())
+
+	By("Register NamespacedCloudProfile controller")
+	Expect((&namespacedcloudprofile.Reconciler{
+		Config: config.NamespacedCloudProfileControllerConfiguration{
+			ConcurrentSyncs: ptr.To(5),
+		},
+	}).AddToManager(ctx, mgr)).To(Succeed())
```
Comment (rfranzke) [inline test/integration/controllermanager/project/project/project_suite_test.go:155]: Maybe you can add a comment why these controllers are added to the integration test of the `Project` controller. I think it's not immediately obvious.

Comment (LucaBernstein) [reply test/integration/controllermanager/project/project/project_suite_test.go:155]: Thank you for the hint. I've added a comment in this commit: 72a1e035c93cd6928784ac8009b4e907f0cda8e9

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**